### PR TITLE
F-32: add virtual router resource

### LINKED ISSUE #32: vRouter integration
**I can't thank you all enough for your hard work on this provider - it's truly invaluable.**
I fully appreciate the decision to focus on VMs and networking, as client workloads generate far more operational effort than managing the cluster itself.   

**Are there any plans to add vRouter support?**   
At the moment I am having to build vRouters using null-resource and firing onevrouter commands, which seems more than a little primitive given the elegance of using a TF provider.

**Is there a 'features roadmap' anywhere?**
I see some mention of a roadmap/timeline in https://github.com/terraform-providers/terraform-provider-opennebula/issues/4   so I assume the work has been on the features, not the future? :smiley:

Thank you again, team - this provider really adds so much value to OpenNebula.

### LINKED ISSUE #140: Create VROUTER
Need to create vRouter via terraform.
But the argument **vrouter = "YES"**, from the issue #32 in **opennebula_template** resource doesn't work.
`│ An argument named "VROUTER" is not expected here.`

### PR DESCRIPTION
**Pull request by treywelsh**
Add new virtual router resource and modify related resource: template and virtual machine

### CODE REVIEW

**Comment by treywelsh** [comment]
@perrfect @turb0steve If you want to give it a try and some feedback, here is a basic implementation of the virtual router resource (you'll need to build the provider from this branch).

**Comment by frousselet** [comment]
Hi @treywelsh,

As seen together, I think a better approach is to follow the Sunstone workflow in order to keep the Terraform provider and the GUI consistant:

1 - Create a new Virtual Router template (1 resource)
2 - Instantiate it (1 resource)

```hcl
resource "opennebula_virtual_router_template" "test" {
  name        = "testacc-vr"
  description = "This is an example of virtual router"

  permissions = "642"
  group       = "oneadmin"

  timeout = 5
  lock    = "USE"
  nic {
    network_id = opennebula_virtual_network.test.id
    ip         = "172.16.100.10"
  }
}

resource "opennebula_virtual_router_instance" "test" {
  count = 3
  
  name        = "testacc-vr-instance-${count.index}"
  template_id = opennebula_virtual_router_template.test.id

  # ...
}
```

Also, we might move `opennebula_template` to `opennebula_virtual_machine_template`.

**Comment by treywelsh** [comment]
I realize that it still miss some stuffs:
- externalize NIC ressource from virtual router: it's how we deal with sub resources now
- test to do: vrouter creation with NICs attach, then add an instance and check if it has nic attached
  EDIT: newly created virtual router instance has nic attached, no need of additional code to manage this

**Comment by treywelsh** [comment]
From this comment, I pushed is a full rework of the virtual router resource:

- introduce more resources, here is the full list: `opennebula_virtual_router`, ` opennebula_virtual_router_nic`, `opennebula_virtual_router_instance_template`, `opennebula_virtual_router_instance`
- mutualize as much code as possible with `opennebula_virtual_machine`, `opennebula_virtual_template`
- introduce provider configuration structure: we need to carry more facilities between CRUD methods: we need some locking to avoid conflicts in resources creation/deletion
  Locking facilities helps to prevent errors when `opennebula_virtual_router_instance` and `opennebula_virtual_router_nic` are created simultaneously.

**Comment by treywelsh** [comment]
I may continue to test it a bit during but I think it's enough satisfying for a review. By the way I remove the draft state.
I hope I didn't make too much mistakes in the documentation part.

File: website/docs/r/virtual_router_instance_template.markdown [hunk 1]
```diff
@@ -0,0 +1,95 @@
+---
+layout: "opennebula"
+page_title: "OpenNebula: opennebula_virtual_router_nic"
+sidebar_current: "docs-opennebula-resource-virtual-router"
+description: |-
+  Provides an OpenNebula virtual router resource.
+---
+
+# opennebula_virtual_router_nic
```
Comment (frousselet) [inline website/docs/r/virtual_router_instance_template.markdown]: Hmm... Copy-paste I guess...

**Review by frousselet — CHANGES_REQUESTED** [review]

File: website/docs/r/virtual_router_instance_template.markdown (hunk 1 above)
Comment (treywelsh) [reply website/docs/r/virtual_router_instance_template.markdown]: The whole page was wrong, I reworked it

**Review by frousselet — APPROVED** [review]

**Review by frousselet — APPROVED** [review]

**Comment by frousselet** [comment]
@treywelsh,
Could you add a comment with `terraform apply` and `terraform destroy` outputs please?

**Comment by treywelsh** [comment]
```
resource "opennebula_virtual_router_instance_template" "test" {
  name        = "testacc-vr-template"
  permissions = "642"
  group       = "oneadmin"
  cpu         = "0.5"
  vcpu        = "1"
  memory      = "512"

  context = {
    dns_hostname = "yes"
    network      = "YES"
  }

  graphics {
    keymap = "en-us"
    listen = "0.0.0.0"
    type   = "VNC"
  }

  os {
    arch = "x86_64"
    boot = ""
  }

  tags = {
    env = "prod"
  }
}

resource "opennebula_virtual_network" "network1" {
  name   = "test-net1"
  type   = "bridge"
  bridge = "onebr"
  mtu    = 1500
  ar {
    ar_type = "IP4"
    size    = 12
    ip4     = "172.16.100.130"
  }
  permissions     = "642"
  group           = "oneadmin"
  security_groups = [0]
  clusters        = [0]
}

resource "opennebula_virtual_router_instance" "test" {
  name        = "testacc-vr-virtual-machine"
  group       = "oneadmin"
  permissions = "642"
  memory      = 128
  cpu         = 0.1

  virtual_router_id = opennebula_virtual_router.test.id
}

resource "opennebula_virtual_router" "test" {
  name        = "testacc-vr"
  permissions = "642"
  group       = "oneadmin"

  instance_template_id = opennebula_virtual_router_instance_template.test.id

  tags = {
    customer = "1"
  }
}

```

Plan:
```
Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # opennebula_virtual_network.network1 will be created
  + resource "opennebula_virtual_network" "network1" {
      + automatic_vlan_id = (known after apply)
      + bridge            = "onebr"
      + clusters          = [
          + 0,
        ]
      + description       = (known after apply)
      + dns               = (known after apply)
      + gateway           = (known after apply)
      + gid               = (known after apply)
      + gname             = (known after apply)
      + group             = "oneadmin"
      + guest_mtu         = 1500
      + hold_size         = (known after apply)
      + id                = (known after apply)
      + ip_hold           = (known after apply)
      + mtu               = 1500
      + name              = "test-net1"
      + network_mask      = (known after apply)
      + permissions       = "642"
      + physical_device   = (known after apply)
      + reservation_size  = (known after apply)
      + reservation_vnet  = (known after apply)
      + security_groups   = [
          + 0,
        ]
      + type              = "bridge"
      + uid               = (known after apply)
      + uname             = (known after apply)
      + vlan_id           = (known after apply)

      + ar {
          + ar_type                = "IP4"
          + computed_global_prefix = (known after apply)
          + computed_ip6           = (known after apply)
          + computed_mac           = (known after apply)
          + computed_ula_prefix    = (known after apply)
          + id                     = (known after apply)
          + ip4                    = "172.16.100.130"
          + size                   = 12
        }
    }

  # opennebula_virtual_router.test will be created
  + resource "opennebula_virtual_router" "test" {
      + gid                  = (known after apply)
      + gname                = (known after apply)
      + group                = "oneadmin"
      + id                   = (known after apply)
      + instance_template_id = (known after apply)
      + name                 = "testacc-vr"
      + permissions          = "642"
      + tags                 = {
          + "customer" = "1"
        }
      + timeout              = 3
      + uid                  = (known after apply)
      + uname                = (known after apply)
    }

  # opennebula_virtual_router_instance.test will be created
  + resource "opennebula_virtual_router_instance" "test" {
      + cpu               = 0.1
      + gid               = (known after apply)
      + gname             = (known after apply)
      + group             = "oneadmin"
      + id                = (known after apply)
      + lcmstate          = (known after apply)
      + memory            = 128
      + name              = "testacc-vr-virtual-machine"
      + on_disk_change    = "swap"
      + pending           = false
      + permissions       = "642"
      + state             = (known after apply)
      + template_disk     = (known after apply)
      + timeout           = 3
      + uid               = (known after apply)
      + uname             = (known after apply)
      + vcpu              = (known after apply)
      + virtual_router_id = (known after apply)

      + vmgroup {
          + role       = (known after apply)
          + vmgroup_id = (known after apply)
        }
    }

  # opennebula_virtual_router_instance_template.test will be created
  + resource "opennebula_virtual_router_instance_template" "test" {
      + context     = {
          + "dns_hostname" = "yes"
          + "network"      = "YES"
        }
      + cpu         = 0.5
      + gid         = (known after apply)
      + gname       = (known after apply)
      + group       = "oneadmin"
      + id          = (known after apply)
      + memory      = 512
      + name        = "testacc-vr-template"
      + permissions = "642"
      + reg_time    = (known after apply)
      + tags        = {
          + "env" = "prod"
        }
      + uid         = (known after apply)
      + uname       = (known after apply)
      + vcpu        = 1

      + graphics {
          + keymap = "en-us"
          + listen = "0.0.0.0"
          + port   = (known after apply)
          + type   = "VNC"
        }

      + os {
          + arch = "x86_64"
        }

      + vmgroup {
          + role       = (known after apply)
          + vmgroup_id = (known after apply)
        }
    }

Plan: 4 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

opennebula_virtual_router_instance_template.test: Creating...
opennebula_virtual_network.network1: Creating...
opennebula_virtual_router_instance_template.test: Creation complete after 1s [id=1]
opennebula_virtual_router.test: Creating...
opennebula_virtual_router.test: Creation complete after 0s [id=0]
opennebula_virtual_router_instance.test: Creating...
opennebula_virtual_network.network1: Still creating... [10s elapsed]
opennebula_virtual_router_instance.test: Still creating... [10s elapsed]
opennebula_virtual_network.network1: Creation complete after 11s [id=1]
opennebula_virtual_router_instance.test: Creation complete after 13s [id=0]

Apply complete! Resources: 4 added, 0 changed, 0 destroyed.
```

And destroy:
```
opennebula_virtual_router_instance_template.test: Refreshing state... [id=1]
opennebula_virtual_network.network1: Refreshing state... [id=1]
opennebula_virtual_router.test: Refreshing state... [id=0]
opennebula_virtual_router_instance.test: Refreshing state... [id=0]

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  - destroy

Terraform will perform the following actions:

  # opennebula_virtual_network.network1 will be destroyed
  - resource "opennebula_virtual_network" "network1" {
      - bridge          = "onebr" -> null
      - clusters        = [
          - 0,
        ] -> null
      - gid             = 0 -> null
      - gname           = "oneadmin" -> null
      - group           = "oneadmin" -> null
      - guest_mtu       = 1500 -> null
      - id              = "1" -> null
      - mtu             = 1500 -> null
      - name            = "test-net1" -> null
      - permissions     = "642" -> null
      - security_groups = [
          - 0,
        ] -> null
      - type            = "bridge" -> null
      - uid             = 0 -> null
      - uname           = "oneadmin" -> null

      - ar {
          - ar_type      = "IP4" -> null
          - computed_mac = "02:00:ac:10:64:82" -> null
          - id           = "0" -> null
          - ip4          = "172.16.100.130" -> null
          - size         = 12 -> null
        }
    }

  # opennebula_virtual_router.test will be destroyed
  - resource "opennebula_virtual_router" "test" {
      - gid                  = 0 -> null
      - gname                = "oneadmin" -> null
      - group                = "oneadmin" -> null
      - id                   = "0" -> null
      - instance_template_id = 1 -> null
      - name                 = "testacc-vr" -> null
      - permissions          = "642" -> null
      - tags                 = {
          - "customer" = "1"
        } -> null
      - timeout              = 3 -> null
      - uid                  = 0 -> null
      - uname                = "oneadmin" -> null
    }

  # opennebula_virtual_router_instance.test will be destroyed
  - resource "opennebula_virtual_router_instance" "test" {
      - cpu               = 0.1 -> null
      - gid               = 0 -> null
      - gname             = "oneadmin" -> null
      - group             = "oneadmin" -> null
      - id                = "0" -> null
      - lcmstate          = 3 -> null
      - memory            = 128 -> null
      - name              = "testacc-vr-virtual-machine" -> null
      - on_disk_change    = "swap" -> null
      - pending           = false -> null
      - permissions       = "642" -> null
      - state             = 3 -> null
      - template_disk     = [] -> null
      - timeout           = 3 -> null
      - uid               = 0 -> null
      - uname             = "oneadmin" -> null
      - virtual_router_id = 0 -> null
    }

  # opennebula_virtual_router_instance_template.test will be destroyed
  - resource "opennebula_virtual_router_instance_template" "test" {
      - context     = {
          - "dns_hostname" = "yes"
          - "network"      = "YES"
        } -> null
      - cpu         = 0.5 -> null
      - gid         = 0 -> null
      - gname       = "oneadmin" -> null
      - group       = "oneadmin" -> null
      - id          = "1" -> null
      - memory      = 512 -> null
      - name        = "testacc-vr-template" -> null
      - permissions = "642" -> null
      - reg_time    = 1653294399 -> null
      - tags        = {
          - "env" = "prod"
        } -> null
      - uid         = 0 -> null
      - uname       = "oneadmin" -> null
      - vcpu        = 1 -> null

      - graphics {
          - keymap = "en-us" -> null
          - listen = "0.0.0.0" -> null
          - type   = "VNC" -> null
        }

      - os {
          - arch = "x86_64" -> null
        }
    }

Plan: 0 to add, 0 to change, 4 to destroy.

Do you really want to destroy all resources?
  Terraform will destroy all your managed infrastructure, as shown above.
  There is no undo. Only 'yes' will be accepted to confirm.

  Enter a value: yes

opennebula_virtual_router_instance.test: Destroying... [id=0]
opennebula_virtual_network.network1: Destroying... [id=1]
opennebula_virtual_network.network1: Destruction complete after 0s
opennebula_virtual_router_instance.test: Still destroying... [id=0, 10s elapsed]
opennebula_virtual_router_instance.test: Still destroying... [id=0, 20s elapsed]
opennebula_virtual_router_instance.test: Destruction complete after 20s
opennebula_virtual_router.test: Destroying... [id=0]
opennebula_virtual_router.test: Destruction complete after 0s
opennebula_virtual_router_instance_template.test: Destroying... [id=1]
opennebula_virtual_router_instance_template.test: Destruction complete after 0s

Destroy complete! Resources: 4 destroyed.
```

**Pull request merged**